File: 002-Recursion/015.subsetSum.js
function subsetSum(arr, n, sum) {
  if (n == 0) {
    return sum === 0 ? 1 : 0;
  }
  return subsetSum(arr, n - 1, sum) + subsetSum(arr, n - 1, sum - arr[n - 1]);
}

const arr = [10, 5, 2, 3, 6];
const sum = 8;

const ans = subsetSum(arr, arr.length, sum);
console.log(ans);

File: 002-Recursion/015.subsetSum.py
'''
    Time Complexity: O(2^n)
'''
def subsetSum(arr, n, sum):
    if n == 0:
        if sum == 0:
            return 1
        else:
            return 0

    return subsetSum(arr, n - 1, sum) + subsetSum(arr, n - 1, sum - arr[n-1])


def main():
    arr = [10, 5, 2, 3, 6]
    sum = 8
    ans = subsetSum(arr, len(arr), sum)
    print(ans)

if __name__ == '__main__':
    main()